Playwright: migrate `wp-media-edit-spec` with additional tests

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates the `wp-media-edit-spec` to Playwright and adds new test cases not implemented in the original file.

- test the sidebar navigation to media, filtering media type using tab bar, rotating image and cancel.
- implement the above tests for both Simple and Atomic sites.

#### Testing instructions

TeamCity
- ensure Playwright E2E tests on this branch completes as expected.

Local
- pull the branch
- transpile TypeScript code
- run the migrated spec file with `mocha --config .mocharc_playwright.yml specs/specs-playwright/wp-media__edit-spec.js`

Parent: #53566 
Related to #52849

worldomonation committed:
Hash out the test, and refactor the image edit process and gallery item clicking process.

worldomonation committed:
Rebase.
Migrate the media-edit spec.

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+
+const selectors = {
+	// Navigation tabs
+	navtabs: '.section-nav-tabs',
```
```suggestion
	navTabs: '.section-nav-tabs',
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+	 * @throws {Error} If requested item could not be located in the gallery. Alternatively,
+	 * if the click action failed to select the gallery item.
+	 */
+	async clickOn( { item }: { item: number } ): Promise< void > {
```
I think `clickItem( index )` might describe the intent better here (with index passed directly). Also would alight better with the `clickTab( name )` method.

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+
+		if ( ! element ) {
+			throw new Error(
+				`Requested item number ${ item } is greater than number of items in gallery.`
```
I don't think this error will ever be thrown as the `waitForSelector` throws itself when the element's not found. Safe to remove the whole check!

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+		await element.waitForElementState( 'visible' );
+		await element.click();
+		await element.waitForElementState( 'stable' );
+		const isSelected = await element.evaluate( ( node ) =>
+			node.classList.contains( 'is-selected' )
+		);
```
```suggestion
		const isSelected = await element.getAttribute( 'class' ).includes( 'is-selected' );
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+		await this.page.click( selector );
+		await preview.waitForElementState( 'stable' );
+		const undoButton = await this.page.waitForSelector( selectors.imageEditorResetButton );
+		await undoButton.isEnabled();
```
This is a potential false negative as we will not know whether the button was eventually enabled or not. This will not throw if the button is not enabled, only return a boolean.
```suggestion
		await undoButton.waitForElementState( 'enabled' );
```

WunderBart replied:
I should mention here that we need to be careful when using this check, whether it's the `isEnabled()` or `waitForElementState( 'enabled' )`. It's because i.e. in Gutenberg, we've started moving away from the disabled attribute due to [some accessibility issues](https://github.com/Automattic/wp-calypso/pull/50911#discussion_r595871737). Instead the `disabled` attribute, the `aria-disabled` might be applied which is not covered by mentioned Playwright's methods. The one element that I know is using this convention is the editor's `Publish` button:

<img width="929" alt="Screenshot 2021-06-17 at 17 13 40" src="https://user-images.githubusercontent.com/1451471/122424950-6b282c80-cf8f-11eb-985a-166754737e92.png">

worldomonation replied:
Interesting to know what's happening behind the scenes, thanks for the heads up!

It is possible to check for the aria value using this:
```
const ariaValue = await testUndoButton.getAttribute('aria-disabled');
```

I ran this against the Gutenberg publish button in its disabled state, and the value of `ariaValue` became `true`. So going forward, we can check for `aria-disabled` attribute as this change rolls out.

For now though, the Image Editor's `Reset` button doesn't seem to have the `aria-disabled` attr yet, so I will accept your suggestion for the time being and perhaps explore writing our own `isDisabled` check in `browser-helper` or `browser-manager`?

WunderBart replied:
> So going forward, we can check for aria-disabled attribute as this change rolls out.

IMO, the biggest danger is that when an element is switched to use `aria-disabled`, we might not notice it because to Playwright, the button will always be enabled then. I thought we would be able to avoid it with Playwright, but you're right - we probably need a custom helper for that. Something like this should do the trick:

```js
/**
 * element-helper.js
 */

export async function waitForElementEnabled( elementHandle, options ) {
	return await this.page.waitForFunction( async () => {
		const isEnabled = await elementHandle.isEnabled();
		const isNotAriaDisabled = await elementHandle
			.getAttribute( 'aria-disabled' )
			.then( ( v ) => v !== 'true' );

		return isEnabled && isNotAriaDisabled ? elementHandle : null;
	}, options );
}
```

Maybe we should also create a linter rule that warns about using the native `waitForElementState( 'enabled' )`? What do you think?

worldomonation replied:
Sorry if I was not clear enough.

Playwright does indeed support querying aria attributes. All you need to do to check for `aria-disabled` is this:

`const ariaValue = await testUndoButton.getAttribute('aria-disabled');`

and it will return the appropriate value that is set for that attribute. 

> Maybe we should also create a linter rule that warns about using the native waitForElementState( 'enabled' )? What do you think?

I am inclined to pass on this at the moment, because I do not have a good sense of how quickly the project to migrate from `disabled` to `aria-disabled` is happening, or if there even is a project for this.

In my opinion, if an existing e2e test is broken by a PR where the changeover is taking place, then it should fall under the purview of the engineer making the change to placate the test. If they are unsure, they are free to reach out to the Quality Squad and we are happy to help. But otherwise, I do not want to prescribe what notion of `enabled` should be used at this point in time.

This is assuming the migration to use `aria-disabled` is an ad-hoc one. If there is a concentrated effort to migrate, then I think a rule can possibly make sense, but at the end of the migration.

WunderBart replied:
> In my opinion, if an existing e2e test is broken by a PR where the changeover is taking place, then it should fall under the purview of the engineer making the change to placate the test.

Agreed! Just to be sure I'm clear myself, I meant that in some cases (like with the `undoButton`), if the element has been migrated to use the `aria-disabled` attribute, the test **won't fail**, because `waitForElementState( "enabled" )` will always return true as the `disabled` attribute will not be present at all. This is how we'd potentially end up with a false-negative test result, which could go unnoticed for a very long time.

worldomonation replied:
@WunderBart Okay, I think I understand what you are saying.

I will look into implementing a `ElementHelper.waitForElementEnabled` and use that in this section of the code.

worldomonation replied:
@WunderBart I have implemented a custom function, `ElementHelper.waitForElementEnabled` which will wait for the element to become 'enabled' within a timeout value.

The helper function `ElementHelper.isElementEnabled` performs the job of checking both the disable property and aria-disabled attribute. 

Note, the `waitForFunction` (and other related functions like `evaluate`) runs in the page context and functions must use JS bindings, not Playwright calls.

WunderBart replied:
See https://github.com/Automattic/wp-calypso/pull/53724#discussion_r657113277

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+	async cancelImageEdit(): Promise< void > {
+		const cancelButton = await this.page.waitForSelector( selectors.imageEditorCancelButton );
+		await cancelButton.click();
+	}
```
Should we make sure that the return to the media modal was performed?

worldomonation replied:
Sure, though in the current usage of `cancelImageEdit` in the spec itself, this is the last step.

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+	async clickTab( {
+		name,
+	}: {
+		name: 'All' | 'Images' | 'Documents' | 'Videos' | 'Audio';
```
I think it's OK here to simply pass as a direct attribute, i.e. `clickTab( 'Images' )`.

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+		// Wait for all placeholders to disappear.
+		// Alternatively, waiting for `networkidle` will achieve the same objective
+		// at the cost of much longer resolving time (~20s).
+		await this.page.waitForSelector( '.is-placeholder', { state: 'hidden' } );
```
Is it possible that `.is-placeholder` might be too generic here? Potentially match some other or even multiple elements on a page.

worldomonation replied:
The `.is-placeholder` label only appears when the loading animation is occurring, so I believe this is sufficient. Green local + TeamCity runs show that it works well under real tests.

Ideally, I would wait until page animations are done, but Playwright does not have a `waitForAnimation` method, so that's out. 

I suspect that if I add another line below this and instruct the gallery listing to be `stable` then it should be a 100% guard.

worldomonation replied:
I've added the guard by checking `is gallery stable?`, and the execution logs show that this addition consumes ~30ms on my machine.

```
  pw:api <= page.click succeeded +0ms
  pw:api => page.waitForSelector started +0ms
  pw:api waiting for selector ".is-placeholder" to be hidden +4ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +5ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +171ms
  pw:api   selector resolved to visible <button class="media-library__list-item is-placeholder">…</button> +409ms
  pw:api   selector did not resolve to any element +518ms
  pw:api <= page.waitForSelector succeeded +4ms
  pw:api => elementHandle.waitForElementState started +0ms
  pw:api   waiting for element to be stable +3ms
  pw:api <= elementHandle.waitForElementState succeeded +32ms
```

Note that Playwright is constantly polling and matching on the `.is-placeholder` class as long as the placeholder imagess are showing, until all placeholder images are gone.

WunderBart replied:
Thanks for checking the timing of these actions. Great to see exactly what's happening under the hood. I guess the waiting for the placeholder element to be gone would be sufficient here then! 👍

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+	}: {
+		name: 'All' | 'Images' | 'Documents' | 'Videos' | 'Audio';
+	} ): Promise< void > {
+		await this.page.waitForSelector( selectors.navtabs );
```
```suggestion
		await this.page.waitForSelector( selectors.navTabs );
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,148 @@
+		name: 'All' | 'Images' | 'Documents' | 'Videos' | 'Audio';
+	} ): Promise< void > {
+		await this.page.waitForSelector( selectors.navtabs );
+		await this.page.click( `${ selectors.navtabs } span:has-text("${ name }")` );
```
```suggestion
		await this.page.click( `${ selectors.navTabs } span:has-text("${ name }")` );
```

WunderBart reviewed (changes requested):
Looking good! 💪 Just a few comments from me

worldomonation committed:
Incorporate WunderBart's suggestion on checking if gallery item is selected.

Co-authored-by: Bart Kalisz <bartlomiej.kalisz@gmail.com>

worldomonation committed:
Address PR comments:
- rename navtabs to navTabs.
- use direct attribute with clickTab.
- add guard after clicking the media tab to wait for gallery element to be stable before continuing.

worldomonation committed:
Further changes requested from review:
- split statement for obtaining class attributes from checking if 'is-seleccted' class attribute exists.
- add confirmation that mediaModal is shown after cancelling image edit.
- change parameters supplied to clickTab in the test spec.
- update JSDoc to match changes

worldomonation committed:
Try implementing a custom waitForElementEnabled function that would look at both `disabled` property and `aria-disabled` attribute when determining enabledness.

worldomonation commented on test/e2e/config/default.json:
```diff
@@ -7,6 +7,7 @@
 	"startBrowserTimeoutMS": 60000,
 	"startAppTimeoutMS": 240000,
 	"afterHookTimeoutMS": 60000,
+	"playwrightTimeoutMS": 30000,
```
This mirrors the value used by Playwright internally - 30000ms or 30s as the timeout.

worldomonation committed:
Add mobile path for `clickTab` method.
Parametrize the simple and atomic tests.

WunderBart commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -0,0 +1,74 @@
+ * @param {number} options.timeout Timeout to override the default value.
+ *
+ */
+export async function waitForElementEnabled(
+	page: Page,
+	selector: string,
+	options?: { timeout?: number }
+): Promise< ElementHandle > {
+	const timeout = ( options?.timeout
+		? options?.timeout
+		: config.get( 'playwrightTimeoutMS' ) ) as number;
+	const elementHandle = await page.waitForSelector( selector );
+
+	for ( let i = 0; i <= timeout; i + 1000 ) {
+		const isEnabled = await isElementEnabled( elementHandle );
+		if ( isEnabled ) {
+			return elementHandle;
+		}
+	}
+
+	throw new Error( `Failed to wait for requested selector ${ selector } to be enabled.` );
+}
```
> Note, the waitForFunction (and other related functions like evaluate) runs in the page context and functions must use JS bindings, not Playwright calls.

Ah, good point. Although we can't really iterate `timeout`-times as you did there, as it's not equivalent to the actual time spent on polling the function. We can leave it to native PW calls:

```suggestion
export async function waitForElementEnabled(
	page: Page,
	elementHandle: ElementHandle,
	options?: { timeout?: number }
): Promise< void > {
	await Promise.all( [
		elementHandle.waitForElementState( 'enabled', options ),
		elementHandle.waitForFunction(
			( element ) => element.ariaDisabled !== 'true',
			elementHandle,
			options
		),
	] );
```

WunderBart replied:
This means we can remove the `isElementEnabled` helper, as there won't be a use case for it.

worldomonation replied:
@WunderBart The code snippet will not work, as `ElementHandle` does not have a `waitForFunction` method - that is reserved for `Page` and `Frame` objects, essentially things representing more than 1 element handle.

In any case, the native PW calls in `waitForFunction` will not work, so we'd need to use JS API to check the aria value. Any of the `eval` variations would have this limitation:
https://playwright.dev/docs/core-concepts#execution-contexts-playwright-and-browser

worldomonation replied:
I've updated the code snippet you have suggested and have changed the signature so it is more aligned with how native Playwright calls are.

Please take a look. I am still trying to get a handle around `pageFunction` related methods and my changes here may not do what it should.

WunderBart replied:
See https://github.com/Automattic/wp-calypso/pull/53724#discussion_r657864906

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,153 @@
+	 * @throws {Error} If requested item could not be located in the gallery, or if the click action
+	 * failed to select the gallery item.
+	 */
+	async clickItem( index: number ): Promise< void > {
```
```suggestion
	async selectItem( index: number ): Promise< void > {
```

worldomonation replied:
Done.

worldomonation committed:
Rebase and address code review comments:
- rename clickItem to selectItem.
- refactor waitForElementEnabled function.

WunderBart commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -0,0 +1,40 @@
+	const elementHandle = await page.waitForSelector( selector, options );
+	await Promise.all( [
+		elementHandle.waitForElementState( 'enabled', options ),
+		page.evaluate( ( element: any ) => element.ariaDisabled !== 'true', elementHandle ),
```
Almost there! 😄 We still need to use the `wait` method as it should poll the condition fn, and then either resolve or throw an error if the condition is not met.  With `page` available we can finally do:
```suggestion
		page.waitForFunction( ( element: any ) => element.ariaDisabled !== 'true', elementHandle ),
```

worldomonation replied:
Right yeah, that escaped me 🤕 Fixed with latest push!

worldomonation committed:
Code review comments
- replace page.evaluate with page.waitforFunction

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,152 @@
+		const isSelected = classAttributes.includes( 'is-selected' );
+		if ( ! isSelected ) {
+			throw new Error( `Failed to select requested item number ${ index }` );
+		}
```
We should be able to skip the middle checks here if we wait for the `is-selected` class after the click:
```suggestion
		// Playwright is able to select the nth matching item given a selector.
		// See https://playwright.dev/docs/selectors#pick-n-th-match-from-the-query-result.
		const elementHandle = await this.page.click( `:nth-match(${ selectors.items }, ${ index })` );
		await this.page.waitForFunction(
			( element ) => element.classList.contains( 'is-selected' ),
			elementHandle
		);
```

worldomonation replied:
This change results in the following:

```
page.waitForFunction: Evaluation failed: TypeError: Cannot read property 'classList' of undefined
            at eval (eval at predicate (eval at evaluate (:303:29)), <anonymous>:1:22)
            at predicate (eval at evaluate (:303:29), <anonymous>:5:30)
            at eval (eval at evaluate (:303:29), <anonymous>:18:78)
            at onRaf (<anonymous>:1667:37)
            at <anonymous>:1677:13
            at Object.run (<anonymous>:1735:28)
            at eval (eval at evaluate (:303:29), <anonymous>:1:14)
            at UtilityScript.evaluate (<anonymous>:305:22)
            at UtilityScript.<anonymous> (<anonymous>:1:44)
```

VSCode suggests that `element` is of `void` type, so that probably has something to do with it.

worldomonation replied:
Yeah, so the suggestion will not work as-is because `page.click` returns a `Promise<void>`, not the ElementHandle.

Therefore, it would still be better to keep Line 74 from the original change and modify the suggested change to reference the ElementHandle returned in that line.

WunderBart replied:
Ah, nice catch! 👍

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,152 @@
+	 */
+	async editImage(): Promise< void > {
+		await this.page.click( selectors.editButton );
+		await this.page.waitForSelector( selectors.mediaModalPreview );
```
This check should not be necessary as the next click call will wait for the target element anyway
```suggestion
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,152 @@
+		const preview = await this.page.waitForSelector( selectors.imageEditorCanvas );
+		const selector = `${ selectors.imageEditorToolbarButton } span:text("Rotate")`;
+		await this.page.click( selector );
+		await preview.waitForElementState( 'stable' );
```
The preview's bounding box doesn't seem to change when rotating the image, so this check should not be necessary.
```suggestion
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -0,0 +1,152 @@
+	 */
+	async cancelImageEdit(): Promise< void > {
+		const cancelButton = await this.page.waitForSelector( selectors.imageEditorCancelButton );
+		await cancelButton.click();
```
```suggestion
		await this.page.click( selectors.imageEditorCancelButton );
```

worldomonation committed:
Incorporate code review comments.

worldomonation committed:
Retrieve elementHandle as `click` does not return reference to the element handle.

WunderBart reviewed (approved):
Nice, looks good! 🐑 it
